janus/src/message.rs

```diff
@@ -226,6 +232,118 @@ impl Display for Interval {
     }
 }
 
+/// Number of seconds from 1970-01-01 to 2000-01-01.
+#[cfg(feature = "database")]
+const TIME_SEC_CONVERSION: u64 = 946_684_800;
+/// Number of milliseconds per second.
+#[cfg(feature = "database")]
+const USEC_PER_SEC: u64 = 1_000_000;
+
+#[cfg(feature = "database")]
+impl<'a> FromSql<'a> for Interval {
+    fn from_sql(
+        _: &postgres_types::Type,
+        raw: &'a [u8],
+    ) -> Result<Self, Box<dyn std::error::Error + Sync + Send>> {
+        use postgres_protocol::types::{Range::*, RangeBound::*};
+
+        match range_from_sql(raw)? {
+            Empty => Err("Interval cannot represent an empty timestamp range".into()),
+            Nonempty(Inclusive(None), _)
+            | Nonempty(Exclusive(None), _)
+            | Nonempty(_, Inclusive(None))
+            | Nonempty(_, Exclusive(None)) => {
+                Err("Interval cannot represent a timestamp range with a null bound".into())
+            }
+            Nonempty(Unbounded, _) | Nonempty(_, Unbounded) => {
+                Err("Interval cannot represent an unbounded timestamp range".into())
+            }
+            Nonempty(Exclusive(_), _) | Nonempty(_, Inclusive(_)) => {
+                Err("Interval can only represent timestamp ranges that are closed at the start and open at the end".into())
+            }
+            Nonempty(Inclusive(Some(start_raw)), Exclusive(Some(end_raw))) => {
+                // These timestamps represent the number of microseconds before (if negative) or
+                // after (if positive) midnight, 1/1/2000.
+                let start_timestamp = timestamp_from_sql(start_raw)?;
+                let end_timestamp = timestamp_from_sql(end_raw)?;
+
+                // Convert to Unix timestamp, in seconds since midnight 1/1/1970.
+                let negative = start_timestamp < 0;
```

Comment:
This code is a little complex.

What about defining a const `Time` that is "the zero SQL timestamp" (i.e. midnight, Jan 1, 2000). Then convert the SQL timestamps into a `Duration` based on a new `from_microseconds` constructor. Then use `add` or `sub` to apply the duration to the base timestamp -- both of these already use a `checked_` function to avoid overflow.

(Unfortunately, since `Duration` is unsigned, we do have to have the additional complexity of `add` or `sub` vs just creating a negative `Duration` and using `add`.)